staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go

```diff
@@ -97,8 +97,7 @@ var (
 			// This metric is used for verifying api call latencies SLO,
 			// as well as tracking regressions in this aspects.
 			// Thus we customize buckets significantly, to empower both usecases.
-			Buckets: []float64{0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0,
-				1.25, 1.5, 1.75, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5, 6, 7, 8, 9, 10, 15, 20, 25, 30, 40, 50, 60},
+			Buckets:        []float64{0.01, 0.05, 0.1, 0.25, 0.5, 1.0, 2.0, 4.0, 10, 20, 40, 60},
```

Comment:
I don't agree for that - we need much better resolution at lower end for performance monitoring.
I can live with less buckets above 1s, but up to 1s (which is our SLO), we need all the buckets that exists.

/hold

Reply:
What are you doing that takes 1 second?

Reply:
Sorry I could have phrased that better. 
Can you help me to understand the scenario where it is vital to distinguish the 0.3-0.35s operations from the 0.35-0.4s.
What reports do you produce from this data? What actions do you take?

Maybe this should be parameterised, so admins like yourself that really really care can have 38 buckets, while most of the world will be fine with 8?

Reply:
If that is changing from say 0.3 to 0.4 for many metrics, it may mean a regression.

Not that saying that it takes low milliseconds, clearly shows take you focus only on very small clusters. Things like listing nodes are easily taking seconds in large enough clusters:
http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadResponsiveness_Prometheus&Resource=nodes&Scope=cluster&Subresource=&Verb=LIST
But even very simple things like "pod bindings" or "post secrets" in high enough percentiles are taking more than hundreds of milliseconds:
http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadResponsiveness_Prometheus&Resource=pods&Scope=namespace&Subresource=binding&Verb=POST
http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadResponsiveness_Prometheus&Resource=secrets&Scope=namespace&Subresource=&Verb=POST

We need a way to distinguish stuff to detect regressions.

Reply:
@bboreham slide 32 in this talk I gave shows API server performance on a large (~250 node) on-premise cluster with the old bucketing, demonstrating plenty of calls that exceed 1s on the p99: https://hashman.ca/srecon-2019/srecon-2019-ehashman.pdf

Granted, some of that was due to a configuration issue that was causing long periods of blocking, but the API server HTTP timeout is 60s, so having the histogram reflect a smaller range than that doesn't make a lot of sense.

Reply:
@ehashman I have not proposed removing the ability to detect calls exceeding 1s, nor have I proposed reducing the range below 60s.

@wojtek-t I believe a few buckets in the 100-1000ms range would give you what you need.  It's probabalistic - if some calls move from 350ms to 400ms, then some will move from 250ms to 500ms at the same time, so you can still detect a regression.

My PR gives 13 buckets; I don't care whether it's a couple more than that, but I don't see any justification for 20+.

Reply:
I should have been more precise from the beginning. Regressions is not the only thing we would like to catch. Imagine 1s SLO.
If all the calls take 501ms, it's completely different situation than all calls taking 999ms. But from your histogram perspective the situation will be exactly the same.

The buckets as you propose doesn't allow us to monitor what is happening with enough granularity to be able to react on what is happening (at the same time leaving place for regressions, as some highly-desired features have to mean some regression).

I can probably live with removing couple buckets (like 0.35 or 0.45), but definitely not that many as you're proposing.

Reply:
Thanks for explaining.  I pushed another commit improving the resolution between 100ms and 2s, and dropped the 10ms bucket as nobody asked for that recently.

I expect call durations to follow a distribution, e.g. Poisson.

Reply:
@bboreham I was addressing your comment:

> What are you doing that takes 1 second?